pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -73,33 +74,129 @@ pub async fn download_layer_file<'a>(
     // If pageserver crashes the temp file will be deleted on startup and re-downloaded.
     let temp_file_path = path_with_suffix_extension(&local_path, TEMP_DOWNLOAD_EXTENSION);
 
-    let (mut destination_file, bytes_amount) = download_retry(
+    let bytes_amount = download_retry(
         || async {
-            let destination_file = tokio::fs::File::create(&temp_file_path)
-                .await
-                .with_context(|| format!("create a destination file for layer '{temp_file_path}'"))
-                .map_err(DownloadError::Other)?;
-
-            let download = storage.download(&remote_path, cancel).await?;
-
-            let mut destination_file =
-                tokio::io::BufWriter::with_capacity(super::BUFFER_SIZE, destination_file);
-
-            let mut reader = tokio_util::io::StreamReader::new(download.download_stream);
-
-            let bytes_amount = tokio::io::copy_buf(&mut reader, &mut destination_file).await;
-
-            match bytes_amount {
-                Ok(bytes_amount) => {
-                    let destination_file = destination_file.into_inner();
-                    Ok((destination_file, bytes_amount))
+            // download to tmp file & make sure tmp file indoe is durable (including file size!)
+            let res = match crate::virtual_file::io_engine::get() {
+                crate::virtual_file::io_engine::IoEngine::NotSet => panic!("unset"),
+                crate::virtual_file::io_engine::IoEngine::StdFs => {
```

Comment:
Should we move this massive `match` into its own function at least?

Thinking `async fn download_to_tmp_file` or sth

Reply:
perhaps... Thinking of the `download.rs` we have other kinds of download to file which need unifying, like the initdb.tar.zst. So yes, doing this refactoring would help further efforts.

Reply:
refactor: 54009053a0949e87e242b7f19545afc3bd8373e2

are you suggesting I should switch initdb download as well in this PR?